### PR TITLE
fix(curriculum): test css rule by using regex directly on code

### LINKED ISSUE #60083: Build a City Skyline Checks are Funky
### Describe the Issue

On step 75, my code below that was only missing `--` on a variable was marked as failing all checks, not just the last check.

### Affected Page

https://www.freecodecamp.org/learn/full-stack-developer/workshop-city-skyline/step-75

### Your code

```
.fb1c {
  width: 100%;
  height: 80%;
  background: repeating-linear-gradient(
      90deg,
      var(--building-color4),
      var(--building-color4) 10%,
      transparent 10%,
      transparent 15%
    ),
    repeating-linear-gradient(
      var(--building-color4),
      var(--building-color4) 10%,
      var(--window-color4) 10%,
      var(window-color4) 90% //ERROR HERE
    );
}

```

### Expected behavior

Based on what the actual issue with the code was, I would have expected to pass every test except for `3. You should use a second color of --window-color4 from 10% to 90%.`

Instead, the code I wrote failed all tests, which made debugging quite difficult.

### Screenshots

_No response_

### System

- Device: [e.g. iPhone 6, Laptop]
- OS: [e.g. iOS 14, Windows 10, Ubuntu 20.04]
- Browser: [e.g. Chrome, Safari]
- Version: [e.g. 22]

### Additional context

https://discord.com/channels/692816967895220344/718214639669870683/1367085759814832239

### PR DESCRIPTION
**Pull request by l3onhard**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/how-to-open-a-pull-request/).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes either locally on my machine, or GitHub Codespaces.

Closes #60083

### CODE REVIEW

File: curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md [hunk 1]
```diff
@@ -25,19 +25,25 @@ Add a `repeating-linear-gradient` to `.fb1c` below the one that's there; use you
 You should not alter the first `repeating-linear-gradient`.
 
 ```js
-assert.match(new __helpers.CSSHelp(document).getStyle(".fb1c")?.getPropVal('background', true), /repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\)/);
+const fb1bCssBlock = code.match(/\.fb1c\s*?{([\s\S]*?)}/)[1].replaceAll(/\s*/g, '');
+const backgroundPropVal = fb1bCssBlock.match(/background:([\s\S]*?)(;|$)/)[1];
+assert.match(backgroundPropVal, /^repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\)/);
 ```
 
 You should add a `repeating-linear-gradient` with a first color of `--building-color4` from `0%` to `10%`.
```
Comment (l3onhard) [inline curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: Since this issue isn’t about the hint texts, I’m putting this suggestion out there instead of committing it directly. (Does that sound like the right way to do it?)
```suggestion
Below the first `repeating-linear-gradient` you should add a second one. It should start with the color `--building-color4` from `0%` to `10%`.
```

File: curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md [hunk 2]
```diff
@@ -25,19 +25,25 @@ Add a `repeating-linear-gradient` to `.fb1c` below the one that's there; use you
...
 
 ```js
-assert.match(new __helpers.CSSHelp(document).getStyle(".fb1c")?.getPropVal('background', true), /repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\),repeating-linear-gradient\(var(\(--building-color4\)(0%)?,var\(--building-color4\)10%|\(--building-color4\)0%10%)/);
+const fb1bCssBlock = code.match(/\.fb1c\s*?{([\s\S]*?)}/)[1].replaceAll(/\s*/g, '');
+const backgroundPropVal = fb1bCssBlock.match(/background:([\s\S]*?)(;|$)/)[1];
+assert.match(backgroundPropVal, /repeating-linear-gradient\(\S*?\),repeating-linear-gradient\(var\(--building-color4\)(0%)?,var\(--building-color4\)(0%)?10%/);
 ```
 
 You should use a second color of `--window-color4` from `10%` to `90%`.
```
Comment (l3onhard) [inline curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]:
```suggestion
Your second `repeating-linear-gradient` should have a second color of `--window-color4` from `10%` to `90%`.
```

File: curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md (hunk 1 above)
Comment (majestic-owl448) [reply curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: this would be the text that appear when the test fail, it should not have instructions to add things
```suggestion
Below the first `repeating-linear-gradient` you should have a second one. It should start with the color `--building-color4` from `0%` to `10%`.
```

Comment (l3onhard) [reply curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: I am onboard! 
Does that mean I can add the hint text improvements to this PR, even though it’s slightly off-topic?

Comment (majestic-owl448) [reply curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: changing the hint text can be part of updating a test to make it easier on the user

if you think it's off topic for this PR, open an issue to discuss the changes, don't use the PR to discuss changes that would go on an other PR

Comment (l3onhard) [reply curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: Got it! 
For this PR, I’ll include the updated hint text.

File: curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md [hunk 3]
```diff
@@ -25,19 +25,25 @@ Add a `repeating-linear-gradient` to `.fb1c` below the one that's there; use you
 You should not alter the first `repeating-linear-gradient`.
 
 ```js
-assert.match(new __helpers.CSSHelp(document).getStyle(".fb1c")?.getPropVal('background', true), /repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\)/);
+const fb1bCssBlock = code.match(/\.fb1c\s*?{([\s\S]*?)}/)[1].replaceAll(/\s*/g, '');
+const backgroundPropVal = fb1bCssBlock.match(/background:([\s\S]*?)(;|$)/)[1];
+assert.match(backgroundPropVal, /^repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\)/);
```
Comment (majestic-owl448) [inline curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md]: spaces have meaning, adding spaces in the wrong place make the repeating linear gradient syntax not work, this is a bit too permissive

**Review by majestic-owl448 — COMMENTED** [review]
I think there should still be at least one test that checks using `getStyle`, otherwise non-working code could be allowed to pass

**Comment by majestic-owl448** [comment]
are you going to continue working on this @l3onhard ?

**Comment by l3onhard** [comment]
Thanks for the reminder! Will work on it this weekend.

**Comment by l3onhard** [comment]
I've been reconsidering the approach of using regex directly on CSS validation. As pointed out by [majestic-owl448](https://github.com/majestic-owl448) [here](https://github.com/freeCodeCamp/freeCodeCamp/pull/63058#discussion_r2556917997): Writing regex that ensures CSS is syntactically correct requires accounting for complex whitespace rules—where it's allowed but not necessary, where it's required, and where it's forbidden.

Including these considerations would result in regex patterns that are difficult to read and understand, hard to maintain and prone to edge cases and bugs.

Here is an example: 

The following test ignores spacing:

```js
const fb1bCssBlock = code.match(/\.fb1c\s*?{([\s\S]*?)}/)[1].replaceAll(/\s*/g, '');
const backgroundPropVal = fb1bCssBlock.match(/background:([\s\S]*?)(;|$)/)[1];
assert.match(
  backgroundPropVal, 
  /^repeating-linear-gradient\(90deg,var\(--building-color4\)(0%)?,var\(--building-color4\)10%,transparent10%,transparent15%\)/
);
```

The following test accounts of spacing: 

```js
const fb1cCssBlock = code.match(/\.fb1c\s*?{([\s\S]*?)}/)[1];
const backgroundPropVal = fb1cCssBlock.match(/background\s*:([\s\S]*?)(;|$)/)[1];
assert.match(
  backgroundPropVal, 
  /^\s*repeating-linear-gradient\(\s*90deg\s*,\s*var\(\s*--building-color4\s*\)\s*(0%\s*)?,\s*var\(\s*--building-color4\s*\)\s*10%\s*,\s*transparent\s+10%\s*,\s*transparent\s+15%\s*\)/
);
```

This is why I’m beginning to believe [Ajay-2005](https://github.com/Ajay-2005) was right when [suggesting to add the following test while keeping the existing ones unchanged](https://github.com/Ajay-2005/freeCodeCamp/blob/7630f977a063bafba3f0e35b5c589a7be29d20d3/curriculum/challenges/english/25-front-end-development/workshop-city-skyline/5d822fd413a79914d39e9917.md?plain=1#L25C1-L31C1):

````md
You should make sure the background value for `.fb1c` exists and contains no syntax errors.

```js
const background = new __helpers.CSSHelp(document).getStyle(".fb1c")?.getPropVal('background', true);
assert.ok(background);
```
````

**Comment by majestic-owl448** [comment]
if you want to keep removing spacing, you can use an already created helper (https://contribute.freecodecamp.org/curriculum-help/#removing-whitespace). It's fine to have a test that ignore spacing, if there is an other one that checks for the validity of the code. Having multiple levels of checking can make it more clear on where is the failure if there are hints that are appropriately descriptive

**Comment by majestic-owl448** [comment]
> This is why I’m beginning to believe [Ajay-2005](https://github.com/Ajay-2005) was right when [suggesting to add the following test while keeping the existing ones unchanged](https://github.com/Ajay-2005/freeCodeCamp/blob/7630f977a063bafba3f0e35b5c589a7be29d20d3/curriculum/challenges/english/25-front-end-development/workshop-city-skyline/5d822fd413a79914d39e9917.md?plain=1#L25C1-L31C1):

I don't think this approach helps remove the confusion on why the tests are failing

**Comment by l3onhard** [comment]
> I don't think this approach helps remove the confusion on why the tests are failing
> > This is why I’m beginning to believe [Ajay-2005](https://github.com/Ajay-2005) was right when [suggesting to add the following test while keeping the existing ones unchanged](https://github.com/Ajay-2005/freeCodeCamp/blob/7630f977a063bafba3f0e35b5c589a7be29d20d3/curriculum/challenges/english/25-front-end-development/workshop-city-skyline/5d822fd413a79914d39e9917.md?plain=1#L25C1-L31C1):
>>> I think there should still be at least one test that checks using `getStyle`, otherwise non-working code could be allowed to pass

Ajay’s test fails if the `background` property is missing or malformed. This allows the learner to confirm that the `background` property exists and therefore to realise that it must contain a syntax error. 

I’m unsure how else we should assist learners with such mistakes. We could check for specific errors (like a missing hyphen in `--building-colour4`) but that’s labour-intensive. Is there another approach I’m overlooking?

**Comment by l3onhard** [comment]
Here is an example of how the approach burrowed by Ajay (introduced by [fix(curriculum): test validity of background property in step 75](https://github.com/freeCodeCamp/freeCodeCamp/pull/63058/commits/282c6f4a17210a1135884f4acd8065e1ecf72692)) behaves, when a syntax error is present (missing hyphen in a CSS variable):

File: curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md [hunk 4]
```diff
@@ -22,22 +22,47 @@ Add a `repeating-linear-gradient` to `.fb1c` below the one that's there; use you
 
 # --hints--
 
+The `background` property of your `.fb1c` should have a valid value. You should check it for syntax errors.
```
Comment (Dario-DC) [inline curriculum/challenges/english/blocks/workshop-city-skyline/5d822fd413a79914d39e9917.md:25]: It seems this doesn't work as intended. I'm able to pass with invalid css such as
```
.fb1c {
  width: 100%;
  height: 80%;
  background: repeating-linear-gradient(
      90deg,
      var(--building-color4),
      var(--building-color4) 10%,
      transparent 10%,
      transparent 15%
    ),
    repeating-linear-gradient(
      var(--building-color4) 0 %,
      var(--building-color4) 10 %,
      var(--window-color4) 10 %,
      var(--window-color4) 90           %
    )
}
```
